### ISSUE TITLE
OrLink only produces correct query results if defined with a single Link in its outgoing set

### ISSUE THREAD

**Issue by einalex** [opencog/atomspace#2644]
As the title says, `OrLink` produces an empty `SetLink` if defined over more than one condition.

Output:
```
hungry people (OrLink (PresentLink ...))
expected: (SetLink (ConceptNode "me"))
(SetLink (ConceptNode "me"))

hungry or thirsty people (OrLink (PresentLink ...) (PresentLink ...))
expected: (SetLink (ConceptNode "me") (ConceptNode "you"))
(SetLink)
```

Test code:
```
#!/usr/bin/env python

from opencog.type_constructors import *
from opencog.bindlink import execute_atom, evaluate_atom
from opencog.utilities import initialize_opencog
from opencog.atomspace import AtomSpace, types

if __name__ == "__main__":
    atomspace = AtomSpace()
    initialize_opencog(atomspace)
    me = ConceptNode("me")
    you = ConceptNode("you")
    hungry = ConceptNode("hungry")
    thirsty = ConceptNode("thirsty")

    StateLink(me, hungry)
    StateLink(you, thirsty)

    query1 = GetLink(
                TypedVariableLink(VariableNode("someone"), TypeNode("ConceptNode")),
                OrLink(PresentLink(StateLink(VariableNode("someone"), hungry))))

    who_is_hungry1 = execute_atom(atomspace, query1)
    print("\nhungry people (OrLink (PresentLink ...))")
    print('expected: (SetLink (ConceptNode "me"))')
    print(who_is_hungry1)

    query2 = GetLink(
                TypedVariableLink(VariableNode("someone"), TypeNode("ConceptNode")),
                OrLink(PresentLink(StateLink(VariableNode("someone"), hungry)),
                            PresentLink(StateLink(VariableNode("someone"), thirsty))))

    who_is_hungry2 = execute_atom(atomspace, query2)
    print("\nhungry or thirsty people (OrLink (PresentLink ...) (PresentLink ...))")
    print('expected: (SetLink (ConceptNode "me") (ConceptNode "you"))')
    print(who_is_hungry2)

```

**Comment by linas** [comment]
Can you use `ChoiceLink` instead? 

The long-term solution for this is to automatically convert (under the covers, inside the engine) or-link+present-link combinations into choice-links but I'm feeling lazy right now. Mostly because one then has to convert `not or present` into `and absent` which rapidly snowballs into conjunctive-normal-form pathologies, and I've been trying to avoid that ...

**Comment by einalex** [comment]
`ChoiceLink` returns empty `SetLink`s in both cases.

**Comment by linas** [comment]
Make sure you are using `Choice x y z` instead of `or Present x Present y Present z`

**Comment by einalex** [comment]
How do I do this if `x` is more than one things that should be present?

**Comment by linas** [comment]
>  if x is more than one thing

Gack. I don't know.  That's a good question and a real bug.

However, I'm confused by what you wrote, it has no variables in it, and and as written it would trash the state.   Perhaps you want the following:
```
execute_atom(
    SequetionalOr(
        Equal( TrueLink(), Get(picker, called_robot, Variable("x")))
        SequentialAnd(
             Equal( FalseLink(), Get(picker, called_robot, Variable("x"))),
             Equal( FalseLink(), Get(picker, has_crate, Variable("x"))))
         SequentialAnd(
             Equal( FalseLink(), Get(picker, called_robot, Variable("x"))),
             Equal( TrueLink(), Get(picker, has_crate, Variable("x"))),
             Equal( TrueLink(), Get(picker, crate_full, Variable("x"))))))
```
That's the style of a traditional "behavior tree" - ; https://en.wikipedia.org/wiki/Behavior_tree_(artificial_intelligence,_robotics_and_control)

There's another example, different style in https://github.com/opencog/atomspace/blob/master/examples/pattern-matcher/sequence.scm

There's also an example of "in context x if condition y then do z" here: https://github.com/opencog/atomspace/blob/master/examples/pattern-matcher/condition.scm 

All of these are very different styles.

**Comment by einalex** [comment]
Actually it has lots of (python) variables in it. So some of those are then VariableNodes. No State trashing ;)
The versions without OrLink already work like a charm.

This is essentially setting a state under a number of different condition sets.
(If your tummy is grumbling and you haven't eaten for a while or if you asked your partner what's for dinner, we assume you want to eat something.)

**Comment by linas** [comment]
> some of those are then VariableNodes

It would be useful if you could show these expanded out.  It's hard to design and debug without these details.

**Comment by einalex** [comment]
One moment...

```
GetLink(
    TypedVariableLink(VariableNode("picker"), TypeNode("ConceptNode")),
    AndLink(
        PresentLink(picker),
        AbsentLink(StateLink(ListLink(VariableNode("picker"), PredicateNode("movement"), ConceptNode("approaching"))),
        OrLink(
            PresentLink(StateLink(ListLink(VariableNode("picker"), PredicateNode("called robot")), ConceptNode("true"))),
            PresentLink(StateLink(ListLink(VariableNode("picker"), PredicateNode("called robot")), ConceptNode("false")),
                StateLink(ListLink(VariableNode("picker"), PredicateNode("has crate")), ConceptNode("false"))),
            PresentLink(StateLink(ListLink(VariableNode("picker"), PredicateNode("called robot")), ConceptNode("false")),
                StateLink(ListLink(VariableNode("picker"), PredicateNode("has crate")), ConceptNode("true")),
                StateLink(ListLink(VariableNode("picker"), PredicateNode("crate is full")), ConceptNode("true")))
        )
    )
)
```

**Comment by linas** [comment]
There's a parentheis misplacement in the AbsentLink above.

**Comment by einalex** [comment]
yeah, copy and paste mistake, I edited the post.

**Comment by linas** [comment]
I'll try putting together a partial solution today. It should work for you example, but won't be entirely general. Full generality is tricky.

**Comment by einalex** [comment]
Thank you :) I appreciate all your work. I saw you were quite busy with that other bug report.

**Comment by linas** [comment]
Heh. Every few years, I put in some major work, and think to myself, "ah hah! now it's really complete!" and then something else comes up.

**Comment by einalex** [comment]
It's a massive project with way too little unit test coverage..which isn't really a surprise given the expressive might of the system. It's hard to think of all the things that people will come up with, and a lot of overhead work.

**Comment by linas** [comment]
OK, so the variant of this, with ChoiceLink, is fixed in #2664
However, the variant of this, with OrLink currently fails.  I'll try that one next.

**Comment by linas** [comment]
See the unit test in `tests/query/choice-compound.scm`

**Comment by linas** [comment]
Not fixed in #2664, its a work-in-progress. Actually, with one minor change, that pull req works, its just that it fails on a more complicated case in the pattern miner. Fixing that now.

**Comment by linas** [comment]
Fixed in #2671 -- the picker example passes, when written with a ChoiceLink.  The equivalent with an OrLink is still broken.

**Comment by linas** [comment]
Getting the variant with `OrLink` to work will be difficult, and I do not want to do that right now. So unless you can convince me that this is urgent, I want to defer this indefinitely.

**Comment by einalex** [comment]
```python
    atomspace = AtomSpace()

    initialize_opencog(atomspace)
    StateLink(ListLink(ConceptNode("alice"), PredicateNode("hungry")), ConceptNode("TRUE"))
    StateLink(ConceptNode("alice"), ConceptNode("at home"))

    query = GetLink(
                    VariableList(
                        TypedVariableLink(VariableNode("x"), TypeNode("ConceptNode")),
                        TypedVariableLink(VariableNode("y"), TypeNode("ConceptNode"))
                    ),
                    ChoiceLink(
                        PresentLink(
                            StateLink(VariableNode("x"), VariableNode("y")),
                            StateLink(ListLink(VariableNode("x"), PredicateNode("hungry")), ConceptNode("TRUE"))
                        )
                    )
                )
    results = execute_atom(atomspace, query)
```

this gives me the following error:

```
Traceback (most recent call last):
  File "./example5.py", line 121, in <module>
    results = execute_atom(atomspace, query)
  File "execute.pyx", line 14, in opencog.execute.execute_atom
IndexError: map::at
```

Removing the second `StateLink` from the query returns an empty SetLink. That should also be wrong.

I'm not sure what the limitations of using only the ChoiceLinks are.

**Comment by linas** [comment]
Hmm. I need to convert that `map::at` failure into a throw with an error message.

There's a limitation in  the current implementation, and that is that the pattern must have some other, non-choice clauses that will independently ground all of the variables in the choice alternatives.  This should have  been flagged during the static-analysis stage, but wasn't.

There are two reasons for this limitation:

* You example requires chaining together clauses according to what variables they share in common. (so that we start with some variable, and then follow the chain to get the others) That chaining exists, but it was designed to only work for the pattern as a  whole. it needs to be ripped up and reworked so that it can work with individual alternatives in a choice-link/or-link. I'm too lazy to do that right now.
* There's a meta issue: if one of the alternatives has variables X,Y in it, and another alternative has Z in it, then we potentially emerge from the pattern knowing X,Y or knowing Z but not both. This is supposed to be detected during static analysis...but ...

whatevr. The current analyzer is supposed to notice that neither X nor Y had a grounding outside of the choice, and complained.  It also should have noticed that there is just one alternative, and just run with that.  I can fix the one-alternative case, and the failure to detect the problem, but having the generalized chaining is much more work.

Are these limitations something you can live with?

**Comment by linas** [comment]
> There's a limitation in the current implementation, 

OK, turns out removing that limitation was easy. Its fixed in  #2676.  The first time I went to do it, I over-thought the problem.

**Comment by einalex** [comment]
This still gives me a Segfault, sadly without stacktrace.

```python
from opencog.type_constructors import *
from opencog.bindlink import execute_atom, evaluate_atom
from opencog.ure import BackwardChainer, ForwardChainer
from opencog.utilities import initialize_opencog
from opencog.atomspace import AtomSpace, TruthValue

atomspace = AtomSpace()
initialize_opencog(atomspace)

StateLink(ListLink(ConceptNode("alice"), PredicateNode("hungry")), ConceptNode("TRUE"))
StateLink(ListLink(ConceptNode("alice"), PredicateNode("thirsty")), ConceptNode("TRUE"))
StateLink(ListLink(ConceptNode("bob"), PredicateNode("hungry")), ConceptNode("FALSE"))
StateLink(ListLink(ConceptNode("bob"), PredicateNode("thirsty")), ConceptNode("FALSE"))
StateLink(ListLink(ConceptNode("charlie"), PredicateNode("hungry")), ConceptNode("FALSE"))
StateLink(ListLink(ConceptNode("charlie"), PredicateNode("thirsty")), ConceptNode("TRUE"))

StateLink(ConceptNode("charlie"), ConceptNode("at home"))
StateLink(ListLink(ConceptNode("at home"), PredicateNode("is cozy")), ConceptNode("TRUE"))

query = GetLink(VariableList(
                        TypedVariableLink(VariableNode("x"), TypeNode("ConceptNode")),
                        TypedVariableLink(VariableNode("y"), TypeNode("ConceptNode"))
                        ),
                    ChoiceLink(
                        PresentLink(
                            StateLink(VariableNode("x"), VariableNode("y")),
                            StateLink(ListLink(VariableNode("y"), PredicateNode("is cozy")), ConceptNode("TRUE")),
                            StateLink(ListLink(VariableNode("x"), PredicateNode("hungry")), ConceptNode("FALSE")),
                            StateLink(ListLink(VariableNode("x"), PredicateNode("thirsty")), ConceptNode("TRUE"))
                        ),
                        PresentLink(
                            StateLink(ListLink(VariableNode("x"), PredicateNode("hungry")), ConceptNode("FALSE")),
                            StateLink(ListLink(VariableNode("x"), PredicateNode("thirsty")), ConceptNode("FALSE"))
                        ),
                        PresentLink(StateLink(ListLink(VariableNode("x"), PredicateNode("hungry")), ConceptNode("TRUE")))
                    ))
results = execute_atom(atomspace, query)

**Comment by einalex** [comment]
Hm, from a logic point of view, there is no reason why all choices (or rather OR-branches) should have the same set of variables. I think I can work around that by introducing a dummy-fact for everything and just checking for that. But that seems super-hackish.

**Comment by linas** [comment]
> there is no reason why all choices should have the same set of variables.

Yes. And that is why there is no check that was made. However, when you use GetLink, you ask for x and y. if y is never pinned down,. what do you expect to get back?

**Comment by einalex** [comment]
I'm sure this is pretty complicated on the code level and has weird interactions with different quote levels.
In naive dream land, it would be convenient to not define variables at all, but get a list of results containing dynamically determined sets of instantiations.

**Comment by linas** [comment]
Fixed in #2683

>  what do you expect to get back?

After the ix, it returns 
```
(SetLink
  (ListLink
    (ConceptNode "alice")
    (VariableNode "y"))
  (ListLink
    (ConceptNode "charlie")
    (ConceptNode "at home"))
  (ListLink
    (ConceptNode "bob")
    (VariableNode "y")))
```
 which is technically wrong, because y is not a Concept, as the type constraints requested. But since y is never grounded, the type constraints are never checked, and so it passes.

I could leave this as-is, or I could alter this to reject any solutions in which ungrounded variables don't match their type constraints .. which is (almost) the same as rejecting solutions that have ungrounded variables, which is almost the same as removing such terms during static analysis, before running the search. ... what's the right answer?

**Comment by einalex** [comment]
Rejecting solutions and removing terms both seem...weird. They also seem inferior to returning an error message telling the users about which sub-patterns don't contain the required variables.

Returning VariableNodes breaks in cases in which the user queries specifically for VariableNodes (when they indicate that with a TypeNode). Apart from that, it's probably slightly worse than returning a set item that just doesn't include the VariableNode, like this:

```
(SetLink
  (ConceptNode "alice")
  (ListLink
    (ConceptNode "charlie")
    (ConceptNode "at home"))
  (ConceptNode "bob"))
```

**Comment by linas** [comment]
You don't have to define variables .. if you don't, it will crawl through and pull out every variable it finds.  This works ... except without strong typing, you risk unexpected results, and sometimes infinite loops. (because the variables, without type constraints, can match anything, including themselves...)

Also, you can inline variable declarations, you don't have to have them up-front in a separate `VariableList` ... just don't put variable declarations into a ChoiceLink ... that is undefined, because variable declarations are handled during static analysis, before the search is performed.

**Comment by linas** [comment]
>  returning an error message telling the users about which sub-patterns don't contain the required variables.

You mean "throwing an exception to tell the user it's a mal-formed query"?

**Comment by einalex** [comment]
Yes.

**Comment by linas** [comment]
You seem to want it both ways... earlier you complained:

>Hm, from a logic point of view, there is no reason why all choices (or rather OR-branches) should have the same set of variables.
I think I can work around that by introducing a dummy-fact for everything

which is a valid complaint.  I , too, realized a while ago that this might be the case, and so was not strict to enforce this check during the static analysis stage.  But then ... 

> returning an error message telling the users about which sub-patterns don't contain the required variables.

Well, that is a request to be strict, and verify that each choice does contain the same variables.  You can't have it both ways - either there is  a strict check, or there isn't...

**Comment by einalex** [comment]
No, no. I was just responding to all your suggestions. They go essentially in two different directions, either allow this kind of use (returning VariableNodes or returning results without VariableNodes) or not allow this kind of use (the three variants we collected for this behavior).

I'm definitely pro allowing it, if possible.

**Comment by linas** [comment]
> I'm definitely pro allowing it, if possible.

Well, its clearly allowed, right now :-) But its a loose end -- its just not clear what the best answer is, what the right thing to do should be -- and I don't mind leaving it unresolved, for a while. What usually happens is that some other, unanticipated example shows up, that places the issue into bright contrast, and makes it clear what should be done.  But sometimes it takes years for someone to unwittingly generate that example.

**Comment by einalex** [comment]
> I guess no `Or`-capabilities for outgoing sets that contain both PresentLinks and non-PresentLinks. That seems to be a rather inconvenient limitation.

OK, so far it's working fine with only StateLinks in the PresentLinks. My problem now is, that I can't add evaluatable conditions like a GreaterThanLink to the individual choices. I think that is one of the differences to the OrLink, and as I said earlier, a rather inconvenient limitation.
Is there a way around it?

**Comment by linas** [comment]
> GreaterThanLink to the individual choices

That's on the to-do list. Some of the big pull reqs of the last few days, such as #2677 lay the groundwork for this, and a path is spelled out in #2673 There's still a fair bit of work left, and I'm kind-of burnt out coding this; I have some other things I have to attend to and so have to take a breather.

> Is there a way around it?

Well, one "obvious" work-around is to bring everything into disjunctive normal form (DNF) -- basically, yank on the Or's until they are at the very top. It then becomes "pattern search A or pattern search B or patterns search C..."  You mentioned earlier that you were auto-generating these searches from some other format, and so perhaps its not too hard for you do do the DNF yourself.  

Of course, I could also bring things into DNF form during the static-analysis step. I'm loathe to do that for various reasons; the wikipedia article on DNF highlights one of the reasons: there are some pathological cases where there is a combinatorial explosion.  I don't know off-hand if your rule-set contains any of these pathological cases... at any rate, query optimization is a hard problem that I'm mulling over. There must be dozens if not hundreds of blog articles on "query optimization", and the AtomSpace is no different -- it doesn't defy the laws of physics. 

So in the end, it depends on how urgent this is: if it's urgent, you have to do the DNF yourself.  If its not urgent, then wait a week or two... I can't really promise a schedule, there are gremlins lurking in the code... and I really do have to attend to other things I've been putting off too long ...

**Comment by linas** [comment]
Well, I've been trying to lay the ground-work for this, and churned a lot of code, done a lot of cleanup, simplified a lot of stuff. (Its a decade-long evolution of ideas and has accumulated a lot of cruft). So the code base has gotten a lot better.

.. but implementing this remains blisteringly complicated. I'm now closer to being able to do this, but the path forward is still not yet clear in my mind. I've made a good number of false starts.

**Comment by alexander-gabriel** [comment]
```
#!/usr/bin/env python

from opencog.type_constructors import *
from opencog.bindlink import execute_atom, evaluate_atom
from opencog.utilities import initialize_opencog
from opencog.atomspace import AtomSpace, types


if __name__ == "__main__":
    atomspace = AtomSpace()
    initialize_opencog(atomspace)
    me = ConceptNode("me")
    you = ConceptNode("you")
    thirsty = ConceptNode("thirsty")
    cold = PredicateNode('cold')

    StateLink(you, thirsty)

    EvaluationLink(
        cold,
        me
    ).tv = TruthValue(1,1)

    query4 = GetLink(
                VariableList(
                    TypedVariableLink(VariableNode("someone"), TypeNode("ConceptNode"))
                ),
                ChoiceLink(
                    PresentLink(StateLink(VariableNode("someone"), thirsty)),
                    PresentLink(IsTrueLink(
                        EvaluationLink(
                            cold,
                            VariableNode("someone")
                        )
                    ))))
    target = execute_atom(atomspace, query4)
    print('actual:\n'+ str(target))
    print('expected:\n(SetLink\n  (ConceptNode "me")\n  (ConceptNode "you"))')
```
could we get this to work?
statelinks make handling consistency a lot easier, but this keeps me from utilizing values and statelinks in parallel.

**Comment by linas** [comment]
Hi @alexander-gabriel I just spotted your post above; I was .. preoccupied with other things. Yes, the above should work. Yes, I will try to make it work "real soon now". I think I even know what the problem is.  It's this:

The query engine has half a dozen different strategies it applies, to find the starting point for a search. This works great, when the search is yoked together with a giant `And` - pick a starting point, and of you go. This breaks down with an `Or` -- different branches of the `Or` may need to start the search in different, unrelated places. Which is what your demo shows.

I think the fix is simple, but as always ... it will take a little bit of time.

(The fix is simple, when the `Or` occurs at the top level. Terribly confusing when its nested deep inside of something. I'll fix the obvious case, and ponder the rest.   If you have a case where the `Or` is deeply nested somewhere, that will help firm up my thoughts.)

**Comment by linas** [comment]
FYI, I got it working, assorted bits and minor pieces need to be cleaned up.

**Comment by linas** [comment]
So pull req #2882 fixes the case in https://github.com/opencog/atomspace/issues/2644#issuecomment-946594038 and so is "usable". There's a few other things I found along the way, and am working to fix those. So there will be a bit more to come "soon".

**Comment by linas** [comment]
OK, so after #2883  is merged, you won't need an explicit `Present` in `query4`. The following should work:
```
query4 = GetLink(
                VariableList(
                    TypedVariableLink(
                        VariableNode("someone"),
                        TypeNode("ConceptNode")
                    )
                ),
                OrLink(
                    PresentLink(
                        StateLink(
                            VariableNode("someone"),
                            thirsty
                        )
                    ),
                    IsTrueLink(
                        EvaluationLink(
                           cold,
                            VariableNode("someone")
                        )
                    )
                )
            )
```
Minor disclaimer: I tested only the scheme variant of this; the python variant should work fine.

**Comment by linas** [comment]
Closing. The final pull req #2885 resolves everything left that was discussed above, and everything now works more or less as you wanted it to, @alexander-gabriel Sorry it took so long to fix all of this. The fixes, in retrospect, were not that hard; I guess I just needed to wait for that proverbial "rainy day" to do it.

The final pull req comment may help clarify my position on all this:

At the top-most level, `OrLink` and `ChoiceLink` behave exactly the same way, when the things being search for are objects, (and not truth values). In other words, picking a bunch of things off the menu (`ChoiceLink`) and making a set union out of them is the same thing as just assuming that `OrLink` applied to a bunch of stuff means 'set-union'. 

(`OrLink` applied to truth values still means 'logical-or' just as before, while `ChoiceLink` applied to truth values doesn't make sense/is undefined.)

Perhaps the prettiest thing to come out of this was that, deep in the bowels, there was already a cartesian product of disjoint parts. Now, right next to it, there is also a sum. So the concepts of both product and sum appear right next to each other, exactly where "the rubber hits the road". So that's both pretty and theoretically reassuring (because in categor/type-theory/logic, the product and sum types are adjoint, and so also it is here, in a certain way.)

**Comment by linas** [comment]
p.s. The biggest blob of code, in https://github.com/opencog/atomspace/issues/2644#issuecomment-640637234 has been transposed to the unit test https://github.com/opencog/atomspace/blob/master/tests/query/or-more-test.scm

p.p.s. If you find more problems, open a new issue. But I'm thinking everything should work well, now.
